corehq/apps/userreports/models.py

```diff
@@ -382,7 +385,7 @@ def named_expression_objects(self):
                 try:
                     named_expressions[name] = ExpressionFactory.from_spec(
                         expression,
-                        FactoryContext(named_expressions=named_expressions, named_filters={})
+                        FactoryContext(named_expressions=named_expressions, named_filters={}, domain=self.domain)
```

Comment:
Similar comment here about pulling the `FactoryContext` creation out of the loop. This one is a bit more complex though since the each successive loop it updates the `named_expressions` dictionary but I think that could be handled by updating the factory context directly.

Reply:
The need for this `while` loop wasn't immediately obvious to me -- but it could just be that I'm slow :upside_down_face: Might it be easier to understand what this is doing if we dropped the `while` loop, and figured out the dependencies first instead? Maybe (building on snopoke's suggestion) ...
```python
named_expressions = {}
factory_context = FactoryContext(
    named_expressions=named_expressions,
    named_filters={}
)
sorted_expressions = sorted_by_dependencies(self.named_expressions)
for name, expression in sorted_expressions:
    # This is not necessary, because `factory_context.named_expressions`
    # is already a reference to `named_expressions`. But it's not
    # obvious, so maybe better to be explicit:
    factory_context.named_expressions = named_expressions
    ...
```
That way, it would be more obvious that it doesn't matter that `named_expressions` isn't fully populated by the time it is used.

I'm happy to consider my suggestion out of scope.

Reply:
The challenge with this approach is that implementing the `sorted_by_dependencies` function would be non-trivial since named expressions could be referenced at any depth within an expression and can reference each other. You would have to build a dependency tree and then sort it topologically.

Reply:
https://github.com/dimagi/commcare-hq/pull/31443/commits/4a986c123fea809e53ac71a7861c6e5fe64ee79a